machine/v1/types_controlplanemachineset.go

```diff
@@ -284,6 +289,17 @@ type GCPFailureDomain struct {
 	Zone string `json:"zone"`
 }
 
+// OpenStackFailureDomain configures failure domain information for the OpenStack platform.
+type OpenStackFailureDomain struct {
+	// ComputeAvailabilityZone is the nova availability zone in which the OpenStack machine provider will create the VM.
+	// +optional
+	ComputeAvailabilityZone string `json:"computeAvailabilityZone,omitempty"`
+
+	// StorageAvailabilityZone is the zone in which the OpenStack machine provider will create the root volume attached to the VM.
+	// +optional
+	StorageAvailabilityZone string `json:"storageAvailabilityZone,omitempty"`
```

Comment:
@stephenfin 👆

Reply:
> I guess the question is, @mdbooth: does it make any sense (or is it even possible) to attach volumes from different Cinder zones, in a storage-flexibility context?

That depends on the value of `[cinder] cross_az_attach` in `nova.conf`. From the [docs](https://docs.openstack.org/nova/latest/configuration/config.html#cinder.cross_az_attach):

> <dl>
> <dt><code>cross_az_attach</code></dt>
> <dd>
> Allow attach between instance and volume in different availability zones.
> 
> If False, volumes attached to an instance must be in the same availability zone in Cinder as the instance availability zone in Nova.
> 
> This also means care should be taken when booting an instance from a volume where source is not “volume” because Nova will attempt to create a volume using the same availability zone as what is assigned to the instance.
> 
> If that AZ is not in Cinder (or `allow_availability_zone_fallback=False` in cinder.conf), the volume create request will fail and the instance will fail the build request.
> 
> By default there is no availability zone restriction on volume attach.
> </dd>
> </dl>

Since that defaults to `False`, it is possible. I don't know if it would make _sense_ though.

Reply:
Also, reproducing from Slack, which covered a different area:

>Speaking with my Nova hat on, we'd usually expect to see `set(compute_azs) == set(volume_azs)` (with the idea being that an AZ would correspond to e.g. a rack). iirc, this is encoded into nova and doing a boot from volume request with a compute AZ of e.g. foo will result in nova requesting a cinder volume in the foo volume AZ.
>
> Couple of caveats, however. In environments with Ceph but without DCN, it's apparently common enough to have multiple nova AZs but a single cinder AZ. Cinder also provides a mechanism to ignore invalid AZ requests so you'd need to make sure that what you request is what you get.

The encoded assumption I referred to is documented [here](https://docs.openstack.org/nova/latest/admin/availability-zones.html#implications-for-moving-servers):

> With respect to availability zones, a server is restricted to a zone if:
>
> 1. The server was created in a specific zone with the `POST /servers` request containing the `availability_zone` parameter.
> 1. If the server create request did not contain the `availability_zone` parameter but the API service is configured for [`default_schedule_zone`](https://docs.openstack.org/nova/latest/configuration/config.html#DEFAULT.default_schedule_zone) then by default the server will be scheduled to that zone.
> 1. The shelved offloaded server was unshelved by specifying the `availability_zone` with the `POST /servers/{server_id}/action` request using microversion 2.77 or greater.
> 1. **[`cinder.cross_az_attach`](https://docs.openstack.org/nova/latest/configuration/config.html#cinder.cross_az_attach) is False, [`default_schedule_zone`](https://docs.openstack.org/nova/latest/configuration/config.html#DEFAULT.default_schedule_zone) is None, the server is created without an explicit zone but with pre-existing volume block device mappings. In that case the server will be created in the same zone as the volume(s) if the volume zone is not the same as [`default_availability_zone`](https://docs.openstack.org/nova/latest/configuration/config.html#DEFAULT.default_availability_zone). See [Resource affinity](https://docs.openstack.org/nova/latest/admin/availability-zones.html#resource-affinity) for details.**

(emphasis mine)

Reply:
In our out-of-band discussion on the matter, we decided to use a general `storageAvailabilityZone` that initially applies to the root volume, and that may just be the default for further attached volumes.

Essentially: even though additional volumes, in future feature developments, might need to be placed in separate availability zones, we deem `storageAvailabilityZone` to have the highest potential of being compatible with future APIs.

Reply:
https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#use-json-field-names-in-godoc will be tricky to conform to:
```
Ensure that the godoc for a field name matches the JSON name, not the Go name, in Go definitions for API objects
```
@JoelSpeed should we advocate for this convention only for the newly created APIs? Requiring it for new fields for already existing APIs will make a mess without renaming already existing fields.

https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#my-proposed-design-looks-like-an-existing-api-we-have-why-am-i-being-told-that-it-must-be-changed gives yes as an answer. Nevertheless, we should fix the naming everywhere then. Starting 4.14?

Reply:
> @JoelSpeed should we advocate for this convention only for the newly created APIs? Requiring it for new fields for already existing APIs will make a mess without renaming already existing fields.

I would argue this is a newly created API and that this convention should be followed, if it isn't already being followed by machine v1 then that's on me and we should definitely fix it!

I think we should at some point at some programatic linting for this rule and start enforcing it. Since it's a minor documentation change I don't see any harm in updating old APIs to follow that rule

Reply:
I'll look into it in the next patchset. Thank you.